core/aws-core/src/main/java/software/amazon/awssdk/awscore/eventstream/EventStreamAsyncResponseTransformer.java

```diff
@@ -431,6 +431,11 @@ public void request(long l) {
                     @Override
                     public void cancel() {
                         dataSubscription.cancel();
+
+                        // Need to complete the futures, otherwise the downstream subscriber will never
```

Comment:
Hmm not sure about this behavior. Presumably the user called `cancel` so why do they need to be notified via `future`?

Reply:
Yeah, I think we'd still want to complete the future because if customer subscriber fails to complete the future when cancelling the subscription, it will hang forever.

Reply:
My concern is that it would be inconsistent with streaming (not event streaming) API's; do we complete the future for an operation like `GetObject` if the subscription is cancelled?

Reply:
It would depend on the implementation of the `AsyncRresponseTransformer`. In `FileAsyncResponseTransformer`, we do complete the future exceptionally after cancelling but we don't in `ByteArrayAsyncResponseTransformer`. 

I don't think we can expect all custom `AsyncResponseTransformer` to complete the future and we can at least make sure the future doesn't hang if they don't complete the future by themselves.

https://github.com/aws/aws-sdk-java-v2/blob/e95033056ac5daf2da2556b55aee6778b261b45f/core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/FileAsyncResponseTransformer.java#L153-L155

Reply:
Right, but I don't think we defacto always complete the future returned by the operation:

```
CompletableFuture cf = s3.getObject(...);
```

`cf` is not the same future returned by the transformer, and we don't complete it when the subscription is cancelled I believe.

Reply:
Spoke offline. After thinking about this more, I think this is fine given that the customer doesn't have direct access to the future returned from an even streaming operation.